DOCSP-28557: reuse client & connection pool faq (#92)
mongodb/docs-csharp: source/faq.txt

the value of ``MaxConnectionPoolSize``. If ``MaxConnectionPoolSize`` is ``100`` and the
application uses only the primary (the default), then only the primary
connection pool grows and there can be at most ``106`` total connections. If the
application uses a :ref:`read preference <read-preference>` to query the
secondary nodes, their pools also grow and there can be ``306`` total
connections.

To support high numbers of concurrent MongoDB threads
within one process, you can increase ``MaxConnectionPoolSize``.

Additionally, connection pools are rate-limited such that each connection pool
can only create, at maximum, the value of ``MaxConnecting`` connections
in parallel at any time. Any additional request stops waiting until one the
following cases occurs:

- One of the existing threads finishes creating a connection.
- An existing connection is checked back into the pool.
- The driver's ability to reuse existing connections improves due to
  rate-limits on connection creation.

You can set the minimum number of concurrent connections to
each server by using the ``MinConnectionPoolSize`` option, which
defaults to ``0``. The connection pool will be initialized with this

> number of sockets. If errors cause any sockets to close and the the
> total number of sockets (both in use and idle) drops below the minimum,
> the driver opens more sockets until the number reaches the minimum.

Reviewer: I: Grammar and typo ("in-use" is an adjective here and so should be hyphenated)

```suggestion
number of sockets. If errors cause any sockets to close and the
total number of sockets (both in-use and idle) drops below the minimum,
the driver opens more sockets until the number reaches the minimum.
```

You can set the maximum number of milliseconds that a connection can
remain idle in the pool by using the ``MaxConnectionIdleTime`` option.
Once a connection is idle for ``MaxConnectionIdleTime``, the driver

> removes and replaces it. This option defaults to 10 minutes.

Reviewer 2: The driver will not replace connections that reach `MaxConnectionIdleTime` unless the pool size is less than `MinPoolSize`. This allows the driver to create more connections during bursts of activity but then slowly reap connections that are no longer required based on steady-state usage.

   see the :ref:`csharp-issues-help` page for next steps and more
   resources.

.. _csharp-faq-connection-pool:

How Does Connection Pooling Work in the {+driver-short+}?
---------------------------------------------------------

Every ``MongoClient`` instance has a built-in connection pool for each server
in your MongoDB topology. Connection pools open sockets on demand to
support concurrent MongoDB operations in your multi-threaded application.

The maximum size of each connection pool is set by the ``MaxConnectionPoolSize`` option, which
defaults to ``100``. If the number of in-use connections to a server reaches
the value of ``MaxConnectionPoolSize``, the next request to that server will wait
until a connection becomes available. 

The driver has a wait queue that limits the number of threads that can
wait for a connection. The size of the wait queue is determined by the
``WaitQueueMultiple`` option, which defaults to ``5``. The wait queue
size is ``WaitQueueMultiple`` x ``MaxConnectionPoolSize``, which is ``500`` by
default. You can also set the wait queue size by specifying the
``WaitQueueSize`` option, which overrides the other settings. However,

> we do not recommend changing the wait queue size from the default.

Author: I feel like putting things in admonitions in the FAQ feels unnecessary, since this content is an "answer," not a guide or anything. Maybe I'm mistaken in feeling that way -- lmk if you think I should do it anyways.

> we do not recommend changing the wait queue size from the default.

Reviewer: That's fair enough.

Each ``MongoClient`` instance opens two additional sockets per server in
your MongoDB topology for monitoring the server's state.

For example, a client connected to a 3-node replica set opens 6
monitoring sockets. It also opens as many sockets as needed to support
an application's threads on each server, up to
the value of ``MaxConnectionPoolSize``. If ``MaxConnectionPoolSize`` is ``100`` and the
application only uses the primary (the default), then only the primary
connection pool grows and there can be at most ``106`` total connections. If the
application uses a :ref:`read preference <read-preference>` to query the

> secondary nodes, their pools also grow and there can be ``306`` total connections.

Author: My understanding is that there are two additional sockets opened no matter the MongoDB topology -- but the total # of connections available depends both on the read preference and the number of nodes in your set.

Additionally, connection pools are rate-limited such that each connection pool
can only create, at maximum, the value of ``MaxConnecting`` connections

> in parallel at any time. Any additional connection stops waiting in the

Author: No, it is the other way around. `MaxConnectionPoolSize` is the max connections that can exist, `MaxConnecting` is how many connections that can be created at once

> in parallel at any time. Any additional connection stops waiting in the

Reviewer: Gotcha. I updated my suggested edit

> following cases:

Author: does this rephrasing clear anything up?

> following cases:

Reviewer: Here are the things I don't understand:

What is a "request"? 
The request is made by what and to what? 
What does it stop waiting for?

> following cases:

Author: rephrased

- One of the existing threads finishes creating a connection, or
  an existing connection is checked back into the pool.
- The driver's ability to reuse existing connections improves due to
  rate-limits on connection creation.

You can set the minimum number of concurrent connections to
each server with the ``MinConnectionPoolSize`` option, which defaults to ``0``.
The connection pool will be initialized with this number of sockets. If
sockets are closed due to any network errors, causing the total number
of sockets (both in use and idle) to drop below the minimum, more
sockets are opened until the minimum is reached.

You can set the maximum number of milliseconds that a connection can
remain idle in the pool before being removed and replaced with

> the ``MaxConnectionIdleTime`` option, which defaults to 10 minutes.

Reviewer: Ah. What if we just remove the monospace value and say "...for this setting is 10 minutes." ?

> the ``MaxConnectionIdleTime`` option, which defaults to 10 minutes.

Author: there is no monospace value here I think, what are you referring to?

``MongoClient`` also has the ``MaxConnectionLifeTime`` option, which
specifies the length of time, 30 minutes by default, that a connection

> can be pooled before expiring.

Author: see prev comment about admonitions

The following default configuration for a ``MongoClient`` works for most
applications:

.. code-block:: csharp
   
   var client = new MongoClient("<connection string>");

Create a client once for each process, and reuse it for all
operations. It is a common mistake to create a new client for each

> request, which is very inefficient.

Reviewer: S: Admonition

Author: see above

To support high numbers of concurrent MongoDB threads

> within one process, you can increase ``MaxConnectionPoolSize``.

Reviewer: S: Should this go up with `MinConnectionPoolSize`? it feels a little tacked on here.

Author: moved

A thread that waits more than the length of time defined by
``MaxConnectionIdleTime`` for a socket raises a connection error. Use this
option if it is more important to bound the duration of operations

> during a load spike than it is to complete every operation.

Reviewer: S: Should this go up with the earlier discussion of `MaxConnectionIdleTime`? Feels tacked on.

Author: Yep. Adapted this content from PyMongo docs, and maybe that content is a little disorganized,

> There is no supported way to terminate a ``MongoClient`` in the driver.

Author: According to James K, [there isn't a way](https://mongodb.slack.com/archives/C03RFE0V9NV/p1680290853739179?thread_ts=1680191758.249449&cid=C03RFE0V9NV)

other settings. However, we do not recommend changing the wait queue
size from the default.

Connection pools are rate-limited. The ``MaxConnecting`` setting
determines the number of connections that the pool can create in
parallel at any time. For example, if the value of ``MaxConnecting`` is
``2``, the third thread that attempts to concurrently check out a

> connection succeeds only when one the following cases occurs:

Reviewer: ```suggestion
connection succeeds only in one of the following cases:
```

You can set the maximum number of milliseconds that a connection can
remain idle in the pool by using the ``MaxConnectionIdleTime`` option.
Once a connection is idle for ``MaxConnectionIdleTime``, the driver
removes it. This option defaults to 10 minutes. If the pool size falls
below ``MinConnectionPoolSize``, the driver removes *and* replaces the
idle connection.

``MongoClient`` also has the ``MaxConnectionLifeTime`` option, which
specifies the length of time, 30 minutes by default, that a connection
can be pooled before expiring.

A thread that waits for a socket for the length of time defined by
the ``ServerSelectionTimeout`` option raises a connection error.
``ServerSelectionTimeout`` defaults to 30 seconds. Use this option if it
is more important to bound the duration of operations during a load

> spike than it is to complete every operation.

Reviewer 2: `ServerSelectionTimeout` will bound how long an operation will wait to find a matching server, but will not bound the total operation time. If you want to bound the total operation time, you should use `maxTimeMS` (exposed on most operations) and/or `socketTimeoutMS` (not recommended). `maxTimeMS` is exposed as `MaxTime` in the various `FindOptions`, `AggregationOptions`, etc. `socketTimeoutMS` is set via the URI or `MongoClientSettings.SocketTimeout`.
